rfq/negotiator_test.go

```diff
@@ -0,0 +1,385 @@
+func assertIncomingSellAcceptTestCase(
+	t *testing.T, tc testCaseIncomingSellAccept) {
+
+	// Create a mock price oracle.
+	mockPriceOracle := &MockPriceOracle{}
+
+	// Register an expected call and response for price oracle method
+	// QueryAskPrice.
+	mockPriceOracle.On(
+		"QueryAskPrice", mock.Anything, mock.Anything,
+		mock.Anything, mock.Anything, mock.Anything,
+	).Return(
+		&OracleResponse{
+			AssetRate: tc.priceOracleAskPrice,
+		}, nil,
+	)
+
+	// Define sell request and sell accept messages.
+	var (
+		assetSpecifier = asset.NewSpecifierFromId(asset.ID{1, 2, 3})
+		peerID         = route.Vertex{1, 2, 3}
+		msgID          = rfqmsg.ID{1, 2, 3}
+	)
+
+	sellRequest := rfqmsg.SellRequest{
+		Peer:           peerID,
+		ID:             msgID,
+		AssetSpecifier: assetSpecifier,
+		PaymentMaxAmt:  lnwire.MilliSatoshi(1000),
+		AssetRateHint:  fn.None[rfqmsg.AssetRate](),
+	}
+
+	sellAccept := rfqmsg.SellAccept{
+		Request:   sellRequest,
+		AssetRate: tc.incomingSellAcceptRate,
+	}
+
+	// Create the negotiator.
+	errChan := make(chan error, 1)
+	negotiator, err := NewNegotiator(NegotiatorCfg{
+		PriceOracle:             mockPriceOracle,
+		OutgoingMessages:        make(chan rfqmsg.OutgoingMsg, 1),
+		AcceptPriceDeviationPpm: tc.acceptPriceDeviationPpm,
+		ErrChan:                 errChan,
+	})
+	require.NoError(t, err)
+
+	// Define the finalise callback function.
+	finalise := func(msg rfqmsg.SellAccept,
+		event fn.Option[InvalidQuoteRespEvent]) {
+
+		// If the actual event is none and the expected status is none,
+		// then we don't need to check anything.
+		if event.IsNone() && tc.quoteRespStatus.IsNone() {
+			return
+		}
+
+		require.Equal(t, tc.quoteRespStatus.IsSome(), event.IsSome())
+
+		// Extract the actual event status.
+		var actualEventStatus QuoteRespStatus
+		event.WhenSome(func(e InvalidQuoteRespEvent) {
+			actualEventStatus = e.Status
+		})
+
+		// Extract the expected event status.
+		var expectedStatus QuoteRespStatus
+		tc.quoteRespStatus.WhenSome(func(e QuoteRespStatus) {
+			expectedStatus = e
+		})
+
+		// Ensure that the actual and expected event statuses are equal.
+		require.Equal(t, expectedStatus, actualEventStatus)
+	}
+
+	// Handle the incoming sell accept message.
+	negotiator.HandleIncomingSellAccept(sellAccept, finalise)
+
+	// Check that there are no errors.
+	select {
+	case err := <-errChan:
+		t.Fatalf("unexpected error: %v", err)
+	default:
+	}
+
+	// Wait for the negotiator to finish.
+	negotiator.Wg.Wait()
+}
+
+// TestHandleIncomingSellAccept tests the handling of an incoming sell accept
+// message.
+func TestHandleIncomingSellAccept(t *testing.T) {
+	defaultQuoteExpiry := time.Now().Add(time.Hour)
+
+	testCases := []testCaseIncomingSellAccept{
+		{
+			name: "accept price just within bounds 1",
+			incomingSellAcceptRate: rfqmsg.AssetRate{
+				Rate:   rfqmath.NewBigIntFixedPoint(1000, 0),
+				Expiry: defaultQuoteExpiry,
+			},
+			priceOracleAskPrice: rfqmsg.AssetRate{
```

Comment:
Nice, seems like we caught some edge cases in the tolerance calculation with these test cases?

Reply:
From my POV, it forced us to be clear about the methodology we use to check tolerance.